scripts/apptainer_generator.py

```diff
@@ -448,35 +495,56 @@ def create_filename(app_root, section_key, actions):
-
-        File naming syntax:
-            <section>_<method>_<action>.sh
-        """
-        # Add your sections here (be sure to modify apptainer/app.def to match)
-        section_meta = {'environment': [],
-                        'post': ['configure', 'make', 'makeinstall'],
-                        'test': ['test']}
-        for section_key, actions in section_meta.items():
-            file_list = self.create_filename(app_root, section_key, actions)
-            for (a_section, file_path) in file_list:
-                with open(file_path, 'r') as section_file:
-                    multi_lines = []
-                    for i, a_line in enumerate(section_file.readlines()):
-                        # next line items need indentation
-                        if i != 0:
-                            multi_lines.append(f'    {a_line}')
-                        else:
-                            multi_lines.append(a_line)
-                    jinja_data[a_section] = ''.join(multi_lines)
+    @staticmethod
+    def add_def_whitespace(content):
+        """
+        Adds whitespace to content for adding to a defintion file
+        """
+        if not content:
+            return content
+        return content.replace('\n', '\n    ')
+
+    def add_definition_includes(self, jinja_data):
+        """
+        Includes files from the repository into the definition file as defined.
+        This is done by setting pre-defined jinja variables for the includes.
+
+        Currently, this is only for the "app" container.
+
+        From syntax (lower, file in repo): apptainer/<container>_<section>_<method>
+        Jinja variable (upper): SECTION_<section>_<method>
+
+        Current supported includes:
+            apptainer/app_environment
+            apptainer/app_post_begin
+            apptainer/app_post_pre_make
+            apptainer/app_post_pre_install
+            apptainer/app_post_post_install
+            apptainer/app_post_end
+            apptainer/app_test_begin
+            apptainer/app_test_end
+        """
+        if self.args.library != 'app':
+            return
+
+        app_name, app_root, _ = Versioner.get_app()
+        sections = ['environment', 'post_begin', 'post_pre_make', 'post_pre_install',
+                    'post_post_install', 'post_end', 'test_begin', 'test_end']
+
+        for section in sections:
+            filename = f'apptainer/app_{section}'
+            full_filename = os.path.join(app_root, filename)
+            if not os.path.isfile(full_filename):
+                continue
+
+            var = f'SECTION_{section.upper()}'
+            file_contents = open(full_filename, 'r').read()
+            contents = f"# Begin include from '{filename}' in {app_name}\n"
+            contents += file_contents
+            contents += f"# End include from '{filename}' in {app_name}\n"
+            contents = self.add_def_whitespace(contents)
+
+            jinja_data[var] = contents
```

Comment:
So if I am reading this right, Application owners can introduce their own scripts? Neat!

Reply:
yeah, I finished that up in conjunction with getting the cardinal container working: https://github.com/loganharbour/cardinal/tree/813_add_context/apptainer